### PR TITLE
Feature/add business logic

### PR DESCRIPTION
**Pull request by ArturMarekNowak**

### CODE REVIEW

File: pkg/database/database.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+		&entities.Income{},
+		&entities.Saving{})
+	if err != nil {
+		panic("Failed to run migrations")
+	}
+}
+
+func OpenConnection() *gorm.DB {
+
+	db, err := gorm.Open(postgres.Open("host=localhost user=postgres password=admin dbname=ExpensesTracker port=5432 sslmode=disable"), &gorm.Config{})
```
Comment (ArturMarekNowak) [inline pkg/database/database.go]: I forgot about using environmental variables

File: pkg/database/entities/expenses_summary.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+	expensesResponse.Savings = make(map[string]float32)
+	for _, element := range expensesSummary.Savings {
+		expensesResponse.Savings[element.Name] = element.Value
+	}
+
+	expensesResponse.Expenses = make(map[string]float32)
+	for _, element := range expensesSummary.Expenses {
+		expensesResponse.Expenses[element.Name] = element.Value
+	}
```
Comment (ArturMarekNowak) [inline pkg/database/entities/expenses_summary.go:69]: I am not sure if this mapping could have been done any better :/ Its simple yet its not exactly a DRY rule though

**Comment by ArturMarekNowak** [comment]
CI/CD is not working also...

File: pkg/api/http/requests/create_expense_summary.go [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+import (
+	"time"
+)
+
+type CreateExpenseSummaryRequest struct {
+	Name                      string
+	ClosedAt                  time.Time
+	UsdToPlnRatio             float64
+	MoneyTransferredToSavings float64
```
Comment (ArturMarekNowak) [inline pkg/api/http/requests/create_expense_summary.go]: float32 is enough

File: pkg/api/http/requests/create_expense_summary.go [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package requests
+
+import (
+	"time"
+)
+
+type CreateExpenseSummaryRequest struct {
+	Name                      string
+	ClosedAt                  time.Time
+	UsdToPlnRatio             float64
```
Comment (ArturMarekNowak) [inline pkg/api/http/requests/create_expense_summary.go]: float32 is enough

File: pkg/api/http/responses/create_expenses_summary.go [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+)
+
+// Source: https://medium.com/@gopal96685/exploring-json-tag-omitempty-in-go-simplify-your-json-output-3ec975585b49
+type CreateExpensesSummaryResponse struct {
+	Id                        uint
+	Name                      string
+	CreatedAt                 time.Time
+	UpdatedAt                 time.Time
+	ClosedAt                  *time.Time `json:",omitempty"`
+	UsdToPlnRatio             float64
```
Comment (ArturMarekNowak) [inline pkg/api/http/responses/create_expenses_summary.go]: float32 is enough

File: pkg/api/http/responses/create_expenses_summary.go [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+	MoneyTransferredToSavings float64
```
Comment (ArturMarekNowak) [inline pkg/api/http/responses/create_expenses_summary.go]: float32 is enough

File: pkg/api/http/responses/get_expenses_summary_response.go [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+import "time"
+
+type GetExpensesSummaryResponse struct {
+	Id                        uint
+	Name                      string
+	CreatedAt                 time.Time
+	UpdatedAt                 time.Time
+	ClosedAt                  *time.Time `json:",omitempty"`
+	UsdToPlnRatio             float64
```
Comment (ArturMarekNowak) [inline pkg/api/http/responses/get_expenses_summary_response.go]: float32 is enough

File: pkg/api/http/responses/get_expenses_summary_response.go [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+	MoneyTransferredToSavings float64
```
Comment (ArturMarekNowak) [inline pkg/api/http/responses/get_expenses_summary_response.go]: float32 is enough

File: pkg/database/entities/expenses_summary.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+	"time"
+)
+
+type ExpensesSummary struct {
+	Id                        uint
+	Name                      string `gorm:"type:character varying(200)"`
+	CreatedAt                 time.Time
+	UpdatedAt                 time.Time
+	ClosedAt                  time.Time `gorm:"type:timestamp with time zone;default:null"`
+	UsdToPlnRatio             float64   `gorm:"type:numeric(4,2)"`
```
Comment (ArturMarekNowak) [inline pkg/database/entities/expenses_summary.go]: float32 is enough

File: pkg/database/entities/expenses_summary.go [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+	MoneyTransferredToSavings float64   `gorm:"type:numeric(10,2)"`
```
Comment (ArturMarekNowak) [inline pkg/database/entities/expenses_summary.go]: float32 is enough

File: pkg/database/entities/expenses_summary.go (hunk 2 above)
Comment (natruszka) [reply pkg/database/entities/expenses_summary.go:69]: Maybe by creating Interface 'IMappable' with function returning pointer to map[string]float32?

File: internal/api/http/services/expenses.go [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"ExpensesTracker/pkg/api/http/requests"
+	"ExpensesTracker/pkg/api/http/responses"
+	"ExpensesTracker/pkg/database"
+	"ExpensesTracker/pkg/database/entities"
+	"errors"
+)
+
+// Source: https://stackoverflow.com/questions/39333102/how-to-create-or-update-a-record-with-gorm
+func UpdateExpense(summaryExpenseId uint, name string, updateSaving requests.CreateExpenseRequest) responses.ExpenseResponse {
```
Comment (natruszka) [inline internal/api/http/services/expenses.go]: I would name it CreateOrUpdateExpense or UpsertExpense

Comment (ArturMarekNowak) [reply internal/api/http/services/expenses.go]: Done

File: pkg/database/entities/expenses_summary.go (hunk 2 above)
Comment (ArturMarekNowak) [reply pkg/database/entities/expenses_summary.go:69]: On second thought I am not entirely sure if it thats possilble. Even if I mark them with an interface I can't define a method that could be invoked on `[]Income` (not `Income`) - similarly `[]Saving` and `[]Expense`. Unfortunately its not supported in Go (I think) which is a pity. First thing that is somewhat similar to this problem that comes to my mind is append function but its not a method, its a simple function with global scope of visibility `func append(slice []Type, elems ...Type) []Type`. Given that I will probably stay with the current solution

Comment (ArturMarekNowak) [reply pkg/database/entities/expenses_summary.go:69]: Another option for interface would be marking `Income`, `Saving` and `Expense` with an interface for common properties but... its also not supported in Go. Only methods

Comment (natruszka) [reply pkg/database/entities/expenses_summary.go:69]: I think we should leave it as it is then. Go is not enough objective language to resolve it with abstraction layer

**Pull request merged**